Sources/BSON/Date+BSONValue.swift

```diff
@@ -2,6 +2,51 @@ import Foundation
 import NIO
 
 extension Date: BSONValue {
+    /*
+     * Initializes a `Date` from ExtendedJSON.
+     *
+     * Parameters:
+     *   - `json`: a `JSON` representing the canonical or relaxed form of ExtendedJSON for a `Date`.
+     *   - `keyPath`: an array of `String`s containing the enclosing JSON keys of the current json being passed in.
+     *              This is used for error messages.
+     *
+     * Returns:
+     *   - `nil` if the provided value does not conform to the `Date` syntax.
+     *
+     * Throws:
+     *   - `DecodingError` if `json` is a partial match or is malformed.
+     */
+    internal init?(fromExtJSON json: JSON, keyPath: [String]) throws {
+        guard let (value, _) = try json.isObjectWithSingleKey(key: "$date", keyPath: keyPath) else {
+            return nil
+        }
+        switch value {
+        case let .object(obj):
+            // canonical extended JSON
+            print(obj)
+            guard let int = try Int64(fromExtJSON: value, keyPath: keyPath + ["$date"]) else {
```

Comment:
so we dont wan't to be adding `$date` to the keypath, as its one of those reserved ones that won't show up in the resultant BSON.